internal/jwtparser/jwt_test.go

```diff
@@ -0,0 +1,203 @@
+	"net/http/httptest"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/juju/errors"
+	jc "github.com/juju/testing/checkers"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/internal/jwtparser"
+)
+
+type jwtParserSuite struct {
+	url        string
+	keySet     jwk.Set
+	signingKey jwk.Key
+	srv        *httptest.Server
+}
+
+var _ = gc.Suite(&jwtParserSuite{})
+
+func (s *jwtParserSuite) SetUpTest(c *gc.C) {
+	keySet, signingKey, err := NewJWKSet()
+	c.Assert(err, jc.ErrorIsNil)
+	s.keySet = keySet
+	s.signingKey = signingKey
+
+	s.srv = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.RequestURI != "/.well-known/jwks.json" {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		hdrs := w.Header()
+		hdrs.Set(`Content-Type`, `application/json`)
+		pub, _ := s.keySet.Key(0)
+		_ = json.NewEncoder(w).Encode(pub)
+	}))
+
+	s.url = s.srv.URL + "/.well-known/jwks.json"
+}
+
+func (s *jwtParserSuite) TearDownTest(_ *gc.C) {
+	s.srv.Close()
+}
+
+func HTTPClient() *http.Client {
+	return &http.Client{
+		Timeout: 5 * time.Second,
+	}
+}
+
+func (s *jwtParserSuite) TestCacheRegistration(c *gc.C) {
+	ctx, done := context.WithCancel(context.Background())
+	defer done()
+	authenticator := jwtparser.NewParserWithHTTPClient(ctx, HTTPClient())
+	err := authenticator.SetJWKSCache(context.Background(), s.url)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *jwtParserSuite) TestCacheRegistrationFailureWithBadURL(c *gc.C) {
+	ctx, done := context.WithCancel(context.Background())
+	defer done()
+	authenticator := jwtparser.NewParserWithHTTPClient(ctx, HTTPClient())
+	err := authenticator.SetJWKSCache(context.Background(), "noexisturl")
+	// We want to make sure that we get an error for a bad url.
+	c.Assert(err, gc.NotNil)
+}
+
+func (s *jwtParserSuite) TestParseJWT(c *gc.C) {
+	ctx, done := context.WithCancel(context.Background())
+	defer done()
+	authenticator := jwtparser.NewParserWithHTTPClient(ctx, HTTPClient())
+	err := authenticator.SetJWKSCache(context.Background(), s.url)
+	c.Assert(err, jc.ErrorIsNil)
+
+	params := JWTParams{
+		audience: "controller-1",
+		subject:  "alice",
+		claims:   map[string]string{"model-1": "read"},
+	}
+	jwt, err := EncodedJWT(params, s.keySet, s.signingKey)
+	c.Assert(err, jc.ErrorIsNil)
+	base64jwt := base64.StdEncoding.EncodeToString(jwt)
+
+	token, err := authenticator.Parse(context.Background(), base64jwt)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(token, gc.NotNil)
+
+	claims := token.PrivateClaims()
+	c.Assert(token.Subject(), gc.Equals, "alice")
+	c.Assert(token.Issuer(), gc.Equals, "test")
+	c.Assert(token.Audience(), jc.DeepEquals, []string{"controller-1"})
+	c.Assert(token.Expiration().After(token.IssuedAt()), gc.Equals, true)
+	c.Assert(claims["access"], jc.DeepEquals, map[string]interface{}{"model-1": "read"})
+}
+
+func (s *jwtParserSuite) TestSimultaneousParse(c *gc.C) {
```

Comment:
What is this test trying to achieve? It is good practice to write a doc description above test justifying its purpose and what it is doing so the test can be groked.

From what I can see the test will finish before the go routines and there is nothing in this design that guarantees any one of these tests will run in parallel to each other.

If we are after a stress test maybe we should consider benchmark or fuz testing.